README.md
# Okanjo Shipping Service Provider

This service takes order information and returns quotes for shipping rates.

## Prerequisites

In order to use this service out-of-the-box, you will need an [EasyPost](http://easypost.com) api key.

## Setup

1. Fork or clone this project.
2. Install dependencies with `npm install`
3. Run `node .` to generate a new `config.js` file. This will generate new encryption and hmac keys. `Control+C` to stop.
4. Update the `config.js` file with your EasyPost api key (test or production)
5. Run `node .` to run the service.
6. Contact support@okanjo.com to register your shipping provider with your Okanjo store or marketplace.

> Note: You may also choose to have Okanjo host your custom shipping provider.

CHANGELOG.md
# Okanjo ShipIt Change Log

When stuff changes, it's described here.

## 2015-06-09
 * Added Raven (Sentry) error reporting mechanism for keeping track of error events
 * Added report function to ShipIt object
 * Added new 400 response "Invalid destination|shipping address" so API can inform customer of bad information
 * Added reference to shipit service on shipping provider to access report function
 * Added reporting to verifying addresses, parcels and creating shipments
 * Added bad address test JSON request to test.php
 * Added warning if EasyPost key is not configured

## 2015-02-26
 * Use a description if already given, or concat carrier and service together

## 2015-02-24
 * Added missing variant string from item quote

## 2015-02-17
 * Cleanup / misc fixes

## 2015-01-08
 * Moved shipping provider to config so it can be easily changed
 * PHP test uses the port from the config file

## 2015-01-07
 * Refactored logic into separate prototype functions to make it easier to extend
 * Added php test (e.g. php test/test.php) to show how to communicate with the service via php

## 2015-01-06
 * Changed the EasyPost rate to return as-is, and we just set some additional minimum properties (id, price, description)

## 2015-01-05
 * Added request logging
 * Added code docs / comments / cleanup

## 2014-12-31
 * Initial 0.1.0 version

test/test.php
<?php



class Cipher {
    private $key, $hmac_key;
    function __construct($key, $hmacKey) {
        $this->key = base64_decode($key);
        $this->hmac_key = base64_decode($hmacKey);
    }

    function secure_random_bytes($len = 16) {

        /*
         * Our primary choice for a cryptographic strong randomness function is
         * openssl_random_pseudo_bytes.
         */
        $SSLstr = '4'; // http://xkcd.com/221/
        if (function_exists('openssl_random_pseudo_bytes') &&
            (version_compare(PHP_VERSION, '5.3.4') >= 0 ||
                substr(PHP_OS, 0, 3) !== 'WIN')
        ) {
            $SSLstr = openssl_random_pseudo_bytes($len, $strong);
            if ($strong) {
                return $SSLstr;
            }
        }

        /*
         * If mcrypt extension is available then we use it to gather entropy from
         * the operating system's PRNG. This is better than reading /dev/urandom
         * directly since it avoids reading larger blocks of data than needed.
         * Older versions of mcrypt_create_iv may be broken or take too much time
         * to finish so we only use this function with PHP 5.3.7 and above.
         * @see https://bugs.php.net/bug.php?id=55169
         */
        if (function_exists('mcrypt_create_iv') &&
            (version_compare(PHP_VERSION, '5.3.7') >= 0 ||
                substr(PHP_OS, 0, 3) !== 'WIN')
        ) {
            $str = mcrypt_create_iv($len, MCRYPT_DEV_URANDOM);
            if ($str !== false) {
                return $str;
            }
        }


        /*
         * No build-in crypto randomness function found. We collect any entropy
         * available in the PHP core PRNGs along with some filesystem info and memory
         * stats. To make this data cryptographically strong we add data either from
         * /dev/urandom or if its unavailable, we gather entropy by measuring the
         * time needed to compute a number of SHA-1 hashes.
         */
        $str = '';
        $bits_per_round = 2; // bits of entropy collected in each clock drift round
        $msec_per_round = 400; // expected running time of each round in microseconds
        $hash_len = 20; // SHA-1 Hash length
        $total = $len; // total bytes of entropy to collect

        $handle = @fopen('/dev/urandom', 'rb');
        if ($handle && function_exists('stream_set_read_buffer')) {
            @stream_set_read_buffer($handle, 0);
        }

        do {
            $bytes = ($total > $hash_len) ? $hash_len : $total;
            $total -= $bytes;

            //collect any entropy available from the PHP system and filesystem
            $entropy = rand() . uniqid(mt_rand(), true) . $SSLstr;
            $entropy .= implode('', @fstat(@fopen(__FILE__, 'r')));
            $entropy .= memory_get_usage() . getmypid();
            $entropy .= serialize($_ENV) . serialize($_SERVER);
            if (function_exists('posix_times')) {
                $entropy .= serialize(posix_times());
            }
            if (function_exists('zend_thread_id')) {
                $entropy .= zend_thread_id();
            }
            if ($handle) {
                $entropy .= @fread($handle, $bytes);
            } else {
                // Measure the time that the operations will take on average
                $c2 = 0;
                $c1 = 0;
                for ($i = 0; $i < 3; $i++) {
                    $c1 = microtime(true);
                    $var = sha1(mt_rand());
                    for ($j = 0; $j < 50; $j++) {
                        $var = sha1($var);
                    }
                    $c2 = microtime(true);
                    $entropy .= $c1 . $c2;
                }

                // Based on the above measurement determine the total rounds
                // in order to bound the total running time.
                $rounds = (int)($msec_per_round * 50 / (int)(($c2 - $c1) * 1000000));

                // Take the additional measurements. On average we can expect
                // at least $bits_per_round bits of entropy from each measurement.
                $iter = $bytes * (int)(ceil(8 / $bits_per_round));
                for ($i = 0; $i < $iter; $i++) {
                    $c1 = microtime();
                    $var = sha1(mt_rand());
                    for ($j = 0; $j < $rounds; $j++) {
                        $var = sha1($var);
                    }
                    $c2 = microtime();
                    $entropy .= $c1 . $c2;
                }

            }
            // We assume sha1 is a deterministic extractor for the $entropy variable.
            $str .= sha1($entropy, true);
        } while ($len > strlen($str));

        if ($handle) {
            @fclose($handle);
        }
        return substr($str, 0, $len);
    }

    function encrypt($text, $iv = null) {

        if (empty($iv)) {
            $iv = function_exists("secure_random_bytes") ? secure_random_bytes(16) : $this->secure_random_bytes(16);
        } else {
            $iv = base64_decode($iv);
        }

        $block = mcrypt_get_block_size(MCRYPT_RIJNDAEL_128, MCRYPT_MODE_CBC);
        $padding = $block - (strlen($text) % $block);
        $text .= str_repeat(chr($padding), $padding);
        $crypttext = mcrypt_encrypt(MCRYPT_RIJNDAEL_128, $this->key, $text, MCRYPT_MODE_CBC, $iv);

        $crypttextenc = base64_encode($crypttext);

        $check = base64_encode(hash_hmac('sha256', $crypttextenc . base64_encode($iv), $this->hmac_key, true));
        $output = $crypttextenc.'$'.base64_encode($iv).'$'.$check;

        return $output;
    }

    function decrypt($encodedString) {

        $parts = explode('$', $encodedString);

        $input = base64_decode($parts[0]);
        $iv = base64_decode($parts[1]);
        $sig = $parts[2];

        $check = base64_encode(hash_hmac('sha256', $parts[0].$parts[1], $this->hmac_key, true));

        if ($check != $sig) {
            error_log('Encrypted Blob has been tampered with...');
            return null;
        }

        $dectext = trim(mcrypt_decrypt(MCRYPT_RIJNDAEL_128, $this->key, $input, MCRYPT_MODE_CBC, $iv), "\0..\32");
        return $dectext;
    }
}


//
// TEST - send the local service a test item blob
//

// Pull the key and hmac_key from the config file
$config = file_get_contents(__DIR__.'/../config.js');

preg_match('/key:\s*(?:\'|")(.+)(?:\'|")/m', $config, $matches);
$key = $matches[1];

preg_match('/hmac_key:\s*(?:\'|")(.+)(?:\'|")/m', $config, $matches);
$hmacKey = $matches[1];


preg_match('/port:\s*([0-9]+)/m', $config, $matches);
$port = $matches[1];


//
// Test the basic informational route
//

$cipher = new Cipher($key, $hmacKey);
$encData = file_get_contents("http://localhost:$port/");
$data = $cipher->decrypt($encData);
echo "$data\n";


//
// Test the calculate route
//

$json = '{"shipping_destination":{"first_name":"John","last_name":"Smith","address_1":"220 E Buffalo St","address_2":"STE 405","city":"Milwaukee","state":"WI","zip":"53202","country":"US","phone":"4148101760"},"shipping_origins":[{"id":"7521","type":"shipping","first_name":"John","last_name":"Smith","address_1":"220 E. Buffalo St","address_2":"Suite 405","city":"Milwaukee","state":"WI","zip":"53210","country":"US","phone":"4148101760"}],"items":[{"product_id":141843,"shipping_methods":[],"get_rates":true,"tax":{"rate":0,"freight_taxable":false,"tax_source":null,"location":null},"quantity":1,"variant":"Model=AOHM-15-FB","product_type":0,"product_donation_perc":null,"product_seller_store_id":5433,"product_brand_id":29,"product_cause_id":null,"product_location_zip":"53406","product_is_local_pickup":0,"product_is_free_shipping":0,"product_use_dynamic_shipping":1,"product_thumbnail_media_id":289507,"product_price":1287,"product_title":"AOHM Series","product_description":"asdf","product_condition":"New","product_parcel":{"length":"16.7","width":"22.5","height":"13.12","weight":"800"},"product_dimensions":{"Model":{"AOHM-10":{"price_modifier":"0"},"AOHM-15":{"price_modifier":"16"},"AOHM-15-FB":{"price_modifier":"60"},"AOHM-15-S":{"price_modifier":"40"},"AOHM-15-S-FB":{"price_modifier":"116"},"AOHM-20":{"price_modifier":"102"},"AOHM-20-FB":{"price_modifier":"178"},"AOHM-20-S-FB":{"price_modifier":"252"},"AOHM-25":{"price_modifier":"315"},"AOHM-25-FB":{"price_modifier":"391"},"AOHM-25-S":{"price_modifier":"389"},"AOHM-25-S-FB":{"price_modifier":"465"},"AOHM-30":{"price_modifier":"584"},"AOHM-30-FB":{"price_modifier":"660"},"AOHM-30-S":{"price_modifier":"658"},"AOHM-30-S-FB":{"price_modifier":"734"},"AOHM-35":{"price_modifier":"811"},"AOHM-35-FB":{"price_modifier":"887"},"AOHM-35-S":{"price_modifier":"885"},"AOHM-35-S-FB":{"price_modifier":"961"},"AOHM-40":{"price_modifier":"1132"},"AOHM-40-FB":{"price_modifier":"1208"},"AOHM-40-S-FB":{"price_modifier":"1282"},"AOHM-5":{"price_modifier":"20"},"AOHM-5-FB":{"price_modifier":"56"},"AOHM-5-S":{"price_modifier":"36"}}},"product_variants":{"Model=AOHM-10":{"stock":"","parcel":{"length":"16.7","width":"22.5","height":"13.12","weight":"800"}},"Model=AOHM-15":{"stock":"","parcel":{"length":"17.09","width":"23.88","height":"15.75","weight":"960"}},"Model=AOHM-15-FB":{"stock":"","parcel":{"length":"17.09","width":"23.88","height":"15.75","weight":"960"}},"Model=AOHM-15-S":{"stock":"","parcel":{"length":"17.09","width":"23.88","height":"15.75","weight":"960"}},"Model=AOHM-15-S-FB":{"stock":"","parcel":{"length":"17.09","width":"23.88","height":"15.75","weight":"960"}},"Model=AOHM-20":{"stock":"","parcel":{"length":"17.09","width":"27.31","height":"18.38","weight":"1200"}},"Model=AOHM-20-FB":{"stock":"","parcel":{"length":"17.09","width":"27.31","height":"18.38","weight":"1200"}},"Model=AOHM-20-S-FB":{"stock":"","parcel":{"length":"17.09","width":"27.31","height":"18.38","weight":"1200"}},"Model=AOHM-25":{"stock":"","parcel":{"length":"17.09","width":"30.18","height":"23.62","weight":"1760"}},"Model=AOHM-25-FB":{"stock":"","parcel":{"length":"17.09","width":"30.18","height":"23.62","weight":"1760"}},"Model=AOHM-25-S":{"stock":"","parcel":{"length":"17.09","width":"30.18","height":"23.62","weight":"1760"}},"Model=AOHM-25-S-FB":{"stock":"","parcel":{"length":"17.09","width":"30.18","height":"23.62","weight":"1760"}},"Model=AOHM-30":{"stock":"","parcel":{"length":"16.7","width":"35.12","height":"27.56","weight":"1920"}},"Model=AOHM-30-FB":{"stock":"","parcel":{"length":"16.7","width":"35.12","height":"27.56","weight":"1920"}},"Model=AOHM-30-S":{"stock":"","parcel":{"length":"16.7","width":"35.12","height":"27.56","weight":"1920"}},"Model=AOHM-30-S-FB":{"stock":"","parcel":{"length":"16.7","width":"35.12","height":"27.56","weight":"1920"}},"Model=AOHM-35":{"stock":"","parcel":{"length":"16.7","width":"37.31","height":"30.19","weight":"2160"}},"Model=AOHM-35-FB":{"stock":"","parcel":{"length":"16.7","width":"37.31","height":"30.19","weight":"2160"}},"Model=AOHM-35-S":{"stock":"","parcel":{"length":"16.7","width":"37.31","height":"30.19","weight":"2160"}},"Model=AOHM-35-S-FB":{"stock":"","parcel":{"length":"16.7","width":"37.31","height":"30.19","weight":"2160"}},"Model=AOHM-40":{"stock":"","parcel":{"length":"16.7","width":"45.12","height":"36.75","weight":"2560"}},"Model=AOHM-40-FB":{"stock":"","parcel":{"length":"16.7","width":"45.12","height":"36.75","weight":"2560"}},"Model=AOHM-40-S-FB":{"stock":"","parcel":{"length":"16.7","width":"45.12","height":"36.75","weight":"2560"}},"Model=AOHM-5":{"stock":"","parcel":{"length":"16.7","width":"18.31","height":"11.81","weight":"560"}},"Model=AOHM-5-FB":{"stock":"","parcel":{"length":"16.7","width":"18.31","height":"11.81","weight":"560"}},"Model=AOHM-5-S":{"stock":"","parcel":{"length":"16.7","width":"18.31","height":"11.81","weight":"560"}}},"product_meta":null,"product_deal_start":null,"product_deal_end":null,"product_promo_start":null,"product_promo_end":null,"product_deal_value":null,"product_media_id_csv":"289507"}]}';

// Bad address test
//$json = '{"shipping_destination":{"first_name":"John","last_name":"Smith","address_1":"Bogus","address_2":"Bogus","city":"Bogus","state":"WI","zip":"53202","country":"US","phone":"4148101760"},"shipping_origins":[{"id":"7521","type":"shipping","first_name":"John","last_name":"Smith","address_1":"220 E. Buffalo St","address_2":"Suite 405","city":"Milwaukee","state":"WI","zip":"53210","country":"US","phone":"4148101760"}],"items":[{"product_id":141843,"shipping_methods":[],"get_rates":true,"tax":{"rate":0,"freight_taxable":false,"tax_source":null,"location":null},"quantity":1,"variant":"Model=AOHM-15-FB","product_type":0,"product_donation_perc":null,"product_seller_store_id":5433,"product_brand_id":29,"product_cause_id":null,"product_location_zip":"53406","product_is_local_pickup":0,"product_is_free_shipping":0,"product_use_dynamic_shipping":1,"product_thumbnail_media_id":289507,"product_price":1287,"product_title":"AOHM Series","product_description":"asdf","product_condition":"New","product_parcel":{"length":"16.7","width":"22.5","height":"13.12","weight":"800"},"product_dimensions":{"Model":{"AOHM-10":{"price_modifier":"0"},"AOHM-15":{"price_modifier":"16"},"AOHM-15-FB":{"price_modifier":"60"},"AOHM-15-S":{"price_modifier":"40"},"AOHM-15-S-FB":{"price_modifier":"116"},"AOHM-20":{"price_modifier":"102"},"AOHM-20-FB":{"price_modifier":"178"},"AOHM-20-S-FB":{"price_modifier":"252"},"AOHM-25":{"price_modifier":"315"},"AOHM-25-FB":{"price_modifier":"391"},"AOHM-25-S":{"price_modifier":"389"},"AOHM-25-S-FB":{"price_modifier":"465"},"AOHM-30":{"price_modifier":"584"},"AOHM-30-FB":{"price_modifier":"660"},"AOHM-30-S":{"price_modifier":"658"},"AOHM-30-S-FB":{"price_modifier":"734"},"AOHM-35":{"price_modifier":"811"},"AOHM-35-FB":{"price_modifier":"887"},"AOHM-35-S":{"price_modifier":"885"},"AOHM-35-S-FB":{"price_modifier":"961"},"AOHM-40":{"price_modifier":"1132"},"AOHM-40-FB":{"price_modifier":"1208"},"AOHM-40-S-FB":{"price_modifier":"1282"},"AOHM-5":{"price_modifier":"20"},"AOHM-5-FB":{"price_modifier":"56"},"AOHM-5-S":{"price_modifier":"36"}}},"product_variants":{"Model=AOHM-10":{"stock":"","parcel":{"length":"16.7","width":"22.5","height":"13.12","weight":"800"}},"Model=AOHM-15":{"stock":"","parcel":{"length":"17.09","width":"23.88","height":"15.75","weight":"960"}},"Model=AOHM-15-FB":{"stock":"","parcel":{"length":"17.09","width":"23.88","height":"15.75","weight":"960"}},"Model=AOHM-15-S":{"stock":"","parcel":{"length":"17.09","width":"23.88","height":"15.75","weight":"960"}},"Model=AOHM-15-S-FB":{"stock":"","parcel":{"length":"17.09","width":"23.88","height":"15.75","weight":"960"}},"Model=AOHM-20":{"stock":"","parcel":{"length":"17.09","width":"27.31","height":"18.38","weight":"1200"}},"Model=AOHM-20-FB":{"stock":"","parcel":{"length":"17.09","width":"27.31","height":"18.38","weight":"1200"}},"Model=AOHM-20-S-FB":{"stock":"","parcel":{"length":"17.09","width":"27.31","height":"18.38","weight":"1200"}},"Model=AOHM-25":{"stock":"","parcel":{"length":"17.09","width":"30.18","height":"23.62","weight":"1760"}},"Model=AOHM-25-FB":{"stock":"","parcel":{"length":"17.09","width":"30.18","height":"23.62","weight":"1760"}},"Model=AOHM-25-S":{"stock":"","parcel":{"length":"17.09","width":"30.18","height":"23.62","weight":"1760"}},"Model=AOHM-25-S-FB":{"stock":"","parcel":{"length":"17.09","width":"30.18","height":"23.62","weight":"1760"}},"Model=AOHM-30":{"stock":"","parcel":{"length":"16.7","width":"35.12","height":"27.56","weight":"1920"}},"Model=AOHM-30-FB":{"stock":"","parcel":{"length":"16.7","width":"35.12","height":"27.56","weight":"1920"}},"Model=AOHM-30-S":{"stock":"","parcel":{"length":"16.7","width":"35.12","height":"27.56","weight":"1920"}},"Model=AOHM-30-S-FB":{"stock":"","parcel":{"length":"16.7","width":"35.12","height":"27.56","weight":"1920"}},"Model=AOHM-35":{"stock":"","parcel":{"length":"16.7","width":"37.31","height":"30.19","weight":"2160"}},"Model=AOHM-35-FB":{"stock":"","parcel":{"length":"16.7","width":"37.31","height":"30.19","weight":"2160"}},"Model=AOHM-35-S":{"stock":"","parcel":{"length":"16.7","width":"37.31","height":"30.19","weight":"2160"}},"Model=AOHM-35-S-FB":{"stock":"","parcel":{"length":"16.7","width":"37.31","height":"30.19","weight":"2160"}},"Model=AOHM-40":{"stock":"","parcel":{"length":"16.7","width":"45.12","height":"36.75","weight":"2560"}},"Model=AOHM-40-FB":{"stock":"","parcel":{"length":"16.7","width":"45.12","height":"36.75","weight":"2560"}},"Model=AOHM-40-S-FB":{"stock":"","parcel":{"length":"16.7","width":"45.12","height":"36.75","weight":"2560"}},"Model=AOHM-5":{"stock":"","parcel":{"length":"16.7","width":"18.31","height":"11.81","weight":"560"}},"Model=AOHM-5-FB":{"stock":"","parcel":{"length":"16.7","width":"18.31","height":"11.81","weight":"560"}},"Model=AOHM-5-S":{"stock":"","parcel":{"length":"16.7","width":"18.31","height":"11.81","weight":"560"}}},"product_meta":null,"product_deal_start":null,"product_deal_end":null,"product_promo_start":null,"product_promo_end":null,"product_deal_value":null,"product_media_id_csv":"289507"}]}';

$uri = "http://localhost:$port/calculate/rates";
$content = $cipher->encrypt($json);
echo $cipher->decrypt($content)."\n";
$http = array('method' => "POST", 'ignore_errors' => true, 'content' => $content);
$headers = array('Content-Type' => "text/plain");
if (!empty($headers)) {

    $headerLines = array();
    foreach($headers as $key => $val) {
        $headerLines[] = "$key: $val";
    }

    $http['header'] = implode("\r\n", $headerLines);
}

$stream_context = stream_context_create(array('http' => $http));
$stream_handle = fopen($uri, 'rb', false, $stream_context);
$stream_metadata = stream_get_meta_data($stream_handle);
$stream_response = @stream_get_contents($stream_handle);
$http_response = $stream_metadata['wrapper_data'][0];
$http_response_array = explode(' ', $http_response);

$response = array("headers" => [], 'contentType' => '', 'data' => $stream_response, "status" => intval($http_response_array[1]));

$arrHeaders = array_splice($stream_metadata['wrapper_data'], 1);
foreach ($arrHeaders as $header) {

    // Add response headers to an associative collection
    $parts = explode(':', $header, 2);
    if (count($parts) == 2) {
        $response['headers'][strtoupper(trim($parts[0]))] = trim($parts[1]);

        // Try to grab the content type
        if (trim(strtolower($parts[0])) == 'content-type')
            $response['contentType'] = trim($parts[1]);
    }
}

$response['data'] = $cipher->decrypt($response['data']);

print_r($response);

index.js
/**
 * Date: 12/17/14 4:26 PM
 *
 * ----
 *
 * (c) Okanjo Partners Inc
 * https://okanjo.com
 * support@okanjo.com
 *
 * https://github.com/okanjo/okanjo-shipit
 *
 * ----
 *
 * TL;DR? see: http://www.tldrlegal.com/license/mit-license
 *
 * The MIT License (MIT)
 * Copyright (c) 2013 Okanjo Partners Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Software, and to permit persons to whom the Software is furnished to do
 * so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

var fs = require('fs'),
    Cipher = require('./lib/cipher');

// If config file exists, use it generate create a new one
if (!fs.existsSync(__dirname+'/config.js')) {

    //
    // GENERATE A NEW CONFIG AND SAVE IT
    //

    // Randomly create a new cipher key and sig
    var randomCipher = new Cipher(),
        template = "module.exports = exports = {\n"+
            "\n"+
            "    /**\n"+
            "     * Encryption protocol\n"+
            "     */\n"+
            "    cipher: {\n"+
            "        key: '"+randomCipher.key.toString('base64')+"',\n"+
            "        hmac_key: '"+randomCipher.hmacKey.toString('base64')+"'\n"+
            "    },\n"+
            "\n"+
            "    /**\n"+
            "     * Port to accept connections on\n"+
            "     */\n"+
            "    port: 54917,\n"+
            "\n"+
            "    /**\n"+
            "     * Easypost API key (http://easypost.com)\n"+
            "     */\n"+
            "    easypost_key: '',\n"+
            "\n"+
            "    /**\n"+
            "     * Shipping provider to use\n"+
            "     */\n"+
            "    provider: require('./lib/shippingprovider'),\n"+
            "\n"+
            "    /**\n"+
            "     * Where to store access logs\n"+
            "     */\n"+
            "    access_log: __dirname + '/access.log',\n"+
            "\n"+
            "    /**\n"+
            "     * Raven/Sentry error reporting URL\n"+
            "     */\n"+
            "    raven_url: null\n"+
            "\n"+
            "};";

    console.log('Generated a new config file!');

    fs.writeFileSync(__dirname+'/config.js', template);
}

var config = require('./config'),
    ShipIt = require('./lib/shipit');

var app = new ShipIt(config);
app.run();

lib/shipit.js
/**
 * Date: 12/18/14 9:10 AM
 *
 * ----
 *
 * (c) Okanjo Partners Inc
 * https://okanjo.com
 * support@okanjo.com
 *
 * https://github.com/okanjo/okanjo-shipit
 *
 * ----
 *
 * TL;DR? see: http://www.tldrlegal.com/license/mit-license
 *
 * The MIT License (MIT)
 * Copyright (c) 2013 Okanjo Partners Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Software, and to permit persons to whom the Software is furnished to do
 * so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

var express = require('express'),
    morgan = require('morgan'),
    fs = require('fs'),
    util = require('util'),
    raven = require('raven'),
    pkg = require('./../package.json'),
    Cipher = require('./cipher'),

    ShipIt = function(config) {

        this.config = config;
        this.cipher = new Cipher(this.config.cipher.key, this.config.cipher.hmac_key);

        // Setup error reporting
        this._initReporting();

        // Instantiate and link the shipping provider
        this.provider = new config.provider(config);
        this.provider.shipit = this;
    };


ShipIt.prototype = {

    constructor: ShipIt,

    /**
     * Starts the server
     */
    run: function run() {

        var app = express(),
            server;

        // Log to a file if desired
        var loggingOptions = {};
        if (this.config.access_log) {
            loggingOptions.stream = fs.createWriteStream(this.config.access_log, {flags: 'a'});
        }
        app.use(morgan('combined', loggingOptions));

        // Bind the routes
        this._bindRoutes(app);

        // Get going
        server = app.listen(this.config.port, function() {
            console.log('%s version %s listening at http://%s:%s', pkg.description, pkg.version, server.address().address, server.address().port);
        });

    },


    /**
     * Initializes error and event reporting, if configured
     * @private
     */
    _initReporting: function() {
        if (typeof this.config.raven_url === "string" && this.config.raven_url.length > 0) {
            this.ravenClient = new raven.Client(this.config.raven_url);
        } else {
            this.ravenClient = null;
        }
    },


    /**
     * Reports an error to the console and via Raven
     */
    report: function() {

        var agg = { err: undefined, meta: { }},
            self = this;

        // Pick the error out and add the rest to the meta array
        for(var i = 0; i < arguments.length; i++) {

            if (typeof arguments[i] === "object" && arguments[i] instanceof Error) {
                console.log(arguments[i].stack);
                agg.err = arguments[i];
            } else {
                console.log(util.inspect(arguments[i], { colors: true }));
                agg.meta['arg'+i] = arguments[i];
            }
        }

        // Report as exception so we always know who was responsible for this
        if (this.ravenClient) {

            // Always get a stack trace
            if (agg.err === undefined) {
                agg.err = new Error('Report: ' + (agg.meta.length > 0 ? agg.meta[0] : "???"));
            }

            this.ravenClient.captureError(agg.err, { extra: agg.meta }, function (res) {
                console.log(' >> Reported as ', self.ravenClient.getIdent(res))
            });
        }
    },


    /**
     * Binds the application routes
     * @param app
     * @private
     */
    _bindRoutes: function(app) {
        var self = this;


        // Pull the raw body out of the request stream
        // so we don't have to use random middle-ware
        app.use(function(req, res, next) {
            var data = '';
            req.setEncoding('utf8');
            req.on('data', function(chunk) {
                data += chunk;
            });
            req.on('end', function() {
                req.rawBody = data;
                next();
            });
        });


        // Show that the service is working and what version
        /**
         * Return the service information. Used for determining whether the service is active and the encryption keys are correct.
         * @param req
         * @param res
         * @private
         */
        app.get('/', function _getAppInfo(req, res) {
            res.setHeader('Content-Type', 'text/plain');
            res.send(self.cipher.encrypt(util.format('%s version %s. Ask your doctor. Use as directed.', pkg.description, pkg.version)));
        });


        // Get the shipping rates
        /**
         * Calculates shipping rates for the given order
         * @param req
         * @param res
         * @private
         */
        app.post('/calculate/rates', function _getShippingRates(req, res) {
            var json = self.cipher.decrypt(req.rawBody);
            if (json) {
                var data = JSON.parse(json);

                //
                // Verify that we have the necessary decision info
                //

                if (!data) {
                    res.status(400).send(self.cipher.encrypt('Bad request'));
                }

                // Verify buyer shipping address is present
                if (!data.hasOwnProperty('shipping_destination')) {
                    res.status(400).send(self.cipher.encrypt('Missing field: shipping_destination'));
                }

                // Verify store shipping addresses are present
                if (!data.hasOwnProperty('shipping_origins') || !data['shipping_origins'].length) {
                    res.status(400).send(self.cipher.encrypt('Missing field: shipping_origins'));
                }

                // Verify there are items to process
                if (!data.hasOwnProperty('items') || !data['items'].length) {
                    res.status(400).send(self.cipher.encrypt('Missing field: items'));
                }

                // Order is good, let's get an estimate on it
                self.provider.calculate(data, function(err, quote) {
                    if (err) {
                        // Allow address errors to funnel back to the API,
                        // so the API can choose to tell the customer about the error in their ways
                        if (typeof err === "string" && err.match(/^Invalid\s([^\s]+)\saddress$/) !== null) {
                            res.status(400).send(self.cipher.encrypt(err));
                        } else {
                            res.status(500).send(self.cipher.encrypt('Service failure'));
                        }
                    } else {
                        res.send(self.cipher.encrypt(JSON.stringify(quote)));
                    }
                });

            } else {
                res.status(400).send(self.cipher.encrypt('Really bad request'));
            }
        });

    }

};


module.exports = exports = ShipIt;